### DISCUSSION TITLE
Troubleshooting Slow File Uploads in Remote MeshCentral Server

### DISCUSSION THREAD

**Discussion by amoljagdalepucsd** [Ylianst/MeshCentral#6265] [Issues] [answered]
Hi Community,

I am running MeshCentral [1.1.22] in hybrid mode.

### Server Setup:

We are running the Remote MeshCentral server behind an Nginx ingress, which is hosted on Azure Kubernetes Service. Ports 443, 4433, and 2121 are accessible on the MeshCentral server. All of this is configured using Kubernetes pod settings. All remote sessions, including remote, backstage, and other functionalities, work fine.

However, the upload file functionality in the files session is extremely slow. I suspected this might be due to the WebRTC configuration. I tried setting the WebRTC property to both `true` and `false` in `config.json`, but this did not significantly affect the upload time.

For example, a 3.28 MB file takes more than 4-5 minutes and sometimes more than 10 minutes to upload. In contrast, when running MeshCentral locally in hybrid mode, it only takes 2-3 seconds to upload the same file. It would be acceptable if it took 15-20 seconds on the remote server.
**filename : meshagent.exe 
size : 3.28 MB**
### Observations:

| No. |              System                                   |        When WebRTC = true              |      When WebRTC = false                 |
|-----|------------------------------------------|-------------------------------------- |-------------------------------------     |
| 1.   |       WIN-SERVER19-Am (Server VM)   |      8 mins                                        |       4 mins 30 seconds, 12 mins        |
| 2.   |       WINSERVER16-Am (Server VM)     |      3 mins, 6 mins                           |       5 mins 45 seconds                       |
| 3.   |      LAP-Anmol-017 (Laptop)                |      5 mins 25 seconds                     |       3 mins 27 seconds                       |

**Is the slow file transfer due to our remote server setup hosted on Azure services with the Nginx ingress controller?**

### config.json

```json
{
    "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
    "settings": {
        "cert": "mydomain.com",
        "sessionSameSite": "lax",
        "port": 4430,
        "aliasPort": 443,
        "agentPort": 2121,
        "agentAliasPort": 443,
        "relayPort": 453,
        "agentPortTls": false,
        "agentPing": 10,
        "browserPing": 10,
        "agentIdleTimeout": 18000,
        "SelfUpdate": false,
        "allowFraming": true,
        "allowLoginToken": true,
        "webRTC": true,
        "tlsoffload": true,
        "agentCoreDump": true,
        "agentLogDump": true,
        "clipboardGet": true,
        "clipboardSet": true,
        "compression": true,
        "wscompression": true,
        "agentwscompression": true,
        "noAgentUpdate": 1,
        "temporaryAgentUpdate": false,
        "agentsInRam": true,
        "desktopMultiplex": true
    },
    "domains": {
        "": {
            "title": "Remote MYmesh",
            "title2": "Mesh",
            "NewAccounts": false,
            "_userNameIsEmail": true,
            "certUrl": "mydomain.com",
            "notifyimage": "MYmesh.png",
            "meshMessengerTitle": "Messenger",
            "nightmode": "userSessionIdleTimeout": 90,
            "allowedorigin": true
        }
    }
}

**Comment by si458** [comment, marked as answer]
file transfers dont get uploaded via webrtc, they only happen over web sockets
webrtc is only used for remote desktop,

also the web ui was transfering the files in tiny bits n bytes
but we did a recent patch to change the bits n bytes size to make it faster
https://github.com/Ylianst/MeshCentral/commit/74d625269964f9d028ee2f537f7d56b1eaceb183

you need either change ur docker image to use the `master` image
OR
install the development branch from npm `npm install Ylianst/MeshCentral`

**Comment by si458** [comment]
you can also try the latest 1.1.25 update and see if the included patch to increase the buffer size for uploads,
speeds up the file upload or not

**Reply by i-am-niru** [reply]
Yes, it speeds up the file upload.
Thanks

**Reply by ahdung** [reply]
this issue re-occured in `1.1.48`, disable webRTC still work, please help.

**Comment by zignaciou** [comment]
Hello,

First of all, thank you for the continued development of MeshCentral.

I would like to request improvements to the file upload mechanism used by MeshCentral.

Problem

In several environments, file uploads are significantly slower than downloads.

For example:

MeshCentral version: 1.2.1
Deployment: Docker
Host: Proxmox VE
Reverse proxy: (if relevant, specify your setup)
Network: Gigabit symmetric connection
Client and server have low CPU and memory utilization

A file of approximately 16 MB takes around 60 seconds to upload, while downloading the same file is almost instantaneous.

Troubleshooting Performed
Verified server resources are not saturated.
Tested with and without WebRTC.
Tested different browsers.
Verified network throughput outside MeshCentral.
No noticeable bottlenecks were found on Docker, Proxmox, or the reverse proxy.

Suggestion

Would it be possible to:

Review the current file upload implementation for performance bottlenecks.
Allow administrators to configure the upload chunk size.
Allow multiple upload chunks to be transferred in parallel.
Provide documentation about expected upload performance and any tunable parameters.

Additional Information

Based on community discussions, other administrators seem to observe similar behavior where uploads are much slower than downloads.

Any clarification regarding current limitations or future plans for file transfer performance would be greatly appreciated.

Upload throughput is approximately 270 KB/s for a 16 MB file on a gigabit network, which suggests the limitation may be related to the transfer protocol implementation rather than network capacity.

Thank you for your work on MeshCentral.

**Reply by si458** [reply]
We are already aware of this, he is a bug with the meshagent with causes slow file transfers and will try looking into it again.

At the moment it's either slow transfers but successfully ones OR fast transfers but they fail 90% of the time

**Reply by zignaciou** [reply]
Thanks for the clarification.
Good to know the issue is already identified in MeshAgent and not related to server configuration.
File uploads are one of the features we use most frequently, so any future improvements in transfer performance would be greatly appreciated, even if introduced as an optional experimental setting.
Thanks again for all the work on MeshCentral.
